MaskedSelect.lua

```diff
@@ -0,0 +1,69 @@
+local MaskedSelect, parent = torch.class('nn.MaskedSelect', 'nn.Module')
+
+--[[ Sets the provided mask value for the module. ]]
+function MaskedSelect:__init()
+  parent.__init(self)
+  self._maskIndices = torch.Tensor()
+  self._maskIndexBuffer = torch.Tensor()
+  self._gradBuffer = torch.Tensor()
+  self._gradMask = torch.Tensor()
+end
+
+--[[ Performs maskedSelect operation. ]]
+function MaskedSelect:updateOutput(input)
+  local input, mask = unpack(input)
+  self.output:maskedSelect(input, mask)
+  return self.output
+end
+
+--[[ Reverse maps unmasked gradOutput back to gradInput. ]]
+function MaskedSelect:updateGradInput(input, gradOutput)
+  local input, mask = unpack(input)
+  self._maskIndexBuffer:range(1, mask:nElement()):resize(mask:size())
+  self._maskIndices:maskedSelect(self._maskIndexBuffer, mask)
+  self._gradBuffer:resize(input:nElement()):zero()
+  self._gradBuffer:scatter(1, self._maskIndices:long(), gradOutput)
```

Comment:
this is not cuda-friendly, as the cuda version of `scatter` requires the indices to be `CudaTensor`.
Also, this line is allocating new memory at every iteration, you can avoid that by having another temporary tensor